src/dmd/mars.d

```diff
@@ -886,6 +888,22 @@ Language changes listed by -transition=id:
             fprintf(global.stdmsg, "semantic3 %s\n", m.toChars());
         m.semantic3(null);
     }
+    if (global.params.includeImports && global.params.link)
+    {
+        // Note: DO NOT USE foreach here because Module.amodules.dim can
+        //       change on each iteration of the loop
+        for (size_t i = 0; i < Module.amodules.dim; i++)
```

Comment:
foreach

Reply:
Did you see the comment above? :)

Reply:
FYI, there wasn't any tests that were failing when this loop was using `foreach` so I added one at "test/compilable/needsmods.d".  The failure is actually the same type of failure as bug 8588, the one where dmd wouldn't discover all imports.  The reason why is because Module.amodules.dim get's "bigger" as you performance semantic analysis on each module, so you need to re-evaluate Module.amodules.dim on each iteration of the loop, otherwise you'll "miss out" on any new modules that were added.